lexicon/providers/yandexcloud.py

```diff
@@ -0,0 +1,307 @@
+        if len(payload["clouds"]) > 1:
+            raise AuthenticationError("Too many clouds found, you must specify one explicitly, or set DNS zone ID")
+
+        return payload["clouds"][0]["id"]
+
+    def _get_folder_id(self, cloud_id: str) -> str:
+        """Gets Folder ID from Resource Manager API
+        https://cloud.yandex.com/en/docs/resource-manager/api-ref/Folder/list
+        """
+        if self._get_provider_option("folder_id"):
+            return self._get_provider_option("folder_id")
+
+        payload: Dict = self._get(f"{self.resources_api_endpoint}/folders", {"cloudId": cloud_id})
+
+        if "folders" not in payload:
+            raise AuthenticationError(f"No folders found for Cloud ID {cloud_id}")
+        if len(payload["folders"]) > 1:
+            raise AuthenticationError("Too many folders found, you must specify one explicitly, or set DNS zone ID")
+
+        return payload["folders"][0]["id"]
+
+    def _get_dns_zone_id(self, folder_id: str) -> str:
+        """Gets Domain ID from DnsZone API
+        https://cloud.yandex.com/en/docs/dns/api-ref/DnsZone/list
+        """
+        payload: Dict = self._get("/zones", {"folderId": folder_id, "filter": f"zone='{self.domain}.'"})
+
+        if "dnsZones" not in payload:
+            raise AuthenticationError(f"No domain found for Folder ID {folder_id}")
+        if len(payload["dnsZones"]) > 1:
+            raise AuthenticationError("Too many domains found, this should not happen")
+
+        return payload["dnsZones"][0]["id"]
+
+    def _list_records(self, rtype=None, name=None, content=None) -> List[Dict]:
+        """List all records. Return an empty list if no records found.
+        type, name and content are used to filter records.
+        https://cloud.yandex.com/en/docs/dns/api-ref/DnsZone/listRecordSets
+        """
+        records: List[Dict] = []
+
+        query_params = {"filter": ""}
+        if rtype:
+            query_params["filter"] = f"type='{rtype}'"
+
+        # name should be in full form, transform it in case it's not
+        if name:
+            name = self._full_name(name) + "."
+            query_params["filter"] += " AND " if query_params["filter"] else ""
+            query_params["filter"] += f"name='{name}'"
+
+        url = f"/zones/{self.domain_id}:listRecordSets"
+        more_pages = True
+        while more_pages:
+            payload: Dict = self._get(url, query_params)
+            if "nextPageToken" in payload:
+                query_params["pageToken"] = payload["nextPageToken"]
+            else:
+                more_pages = False
+
+            # nothing to list
+            if "recordSets" not in payload:
+                break
+
+            for record in payload["recordSets"]:
+                for entry in record["data"]:
+                    # artificially generate ID for deletion
+                    record_id = f"{record['type']}${self._relative_name(record['name'])}${entry}"
```

Comment:
There are two approaches to generate ids for providers that do not provide native ids in the records:
1) Generate a base64 encoding of a string of kind `type/name`
2) Generate a hash of record content (example: https://github.com/AnalogJ/lexicon/blob/bd347b59b0f0ab098ecb8169e1e3609b09a90935/lexicon/providers/godaddy.py#L285-L291)

For 1), I prefer a hash over the string itself because it gives identifiers more transportable, since base64 is primarly designed for that.

For 2), it gives the shortest ids, which has my preference, at the cost that when an identifier of this kind is provider to `_update_record()` or `_delete_record`, you have first to list all records in the zone to recompute the ids and see which one is matching.

Nonetheless I have a slight preference for 2) because of the short ids and the fact that its internals are truly hidden, but it is your call.